File: src/calculate.rs
use crate::models::Position;

pub fn calculate_expected_gain_loss(position: Position, expected_price: f32) -> f32 {
    let cost_basis = position.price * position.amount_owned;
    let mut amount_still_owned = position.amount_owned;

    let mut net_options_premiums: f32 = 0.0;
    let mut net_options_proceedings: f32 = 0.0;
    for b_options in position.options_bought.iter() {
        net_options_premiums -= b_options.premiums();
        if b_options.should_exercise(expected_price) {
            let txn_result = b_options.exercise();
            amount_still_owned += txn_result.amount_change;
            net_options_proceedings += txn_result.cash_change;
        }
    }
    for s_options in position.options_sold.iter() {
        net_options_premiums += s_options.premiums();
        if s_options.should_exercise(expected_price) {
            let txn_result = s_options.exercise();
            amount_still_owned -= txn_result.amount_change;
            net_options_proceedings -= txn_result.cash_change;
        }
    }

    net_options_premiums + net_options_proceedings + amount_still_owned * expected_price
        - cost_basis
}

#[cfg(test)]
mod tests {
    use crate::calculate::calculate_expected_gain_loss;
    use crate::models::*;

    #[test]
    fn no_options() {
        let position = Position {
            price: 15.0,
            amount_owned: 500.0,
            ..Position::default()
        };
        assert_eq!(calculate_expected_gain_loss(position, 16.0), 500.0);
    }

    #[test]
    fn with_worthless_options() {
        let position = Position {
            price: 15.0,
            amount_owned: 100.0,
            options_bought: vec![Option {
                strike_price: 14.5,
                contract_type: OptionType::Put,
                amount: 1.0,
                price: 0.25,
            }],
            options_sold: vec![Option {
                strike_price: 16.0,
                contract_type: OptionType::Call,
                amount: 1.0,
                price: 0.15,
            }],
        };
        assert_eq!(calculate_expected_gain_loss(position, 15.5), 40.0);
    }

    #[test]
    fn with_exercised_sold_call_option() {
        let position = Position {
            price: 15.0,
            amount_owned: 100.0,
            options_bought: vec![Option {
                strike_price: 14.5,
                contract_type: OptionType::Put,
                amount: 1.0,
                price: 0.25,
            }],
            options_sold: vec![Option {
                strike_price: 16.0,
                contract_type: OptionType::Call,
                amount: 1.0,
                price: 0.15,
            }],
        };
        assert_eq!(calculate_expected_gain_loss(position, 16.5), 90.0);
    }

    #[test]
    fn with_exercised_bought_call_option() {
        let position = Position {
            price: 15.0,
            amount_owned: 100.0,
            options_bought: vec![Option {
                strike_price: 14.5,
                contract_type: OptionType::Call,
                amount: 1.0,
                price: 0.25,
            }],
            ..Position::default()
        };
        assert_eq!(calculate_expected_gain_loss(position, 15.5), 125.0);
    }

    #[test]
    fn with_exercised_sold_put_option() {
        let position = Position {
            price: 15.0,
            amount_owned: 100.0,
            options_sold: vec![Option {
                strike_price: 14.5,
                contract_type: OptionType::Put,
                amount: 1.0,
                price: 0.25,
            }],
            ..Position::default()
        };
        assert_eq!(calculate_expected_gain_loss(position, 13.0), -325.0);
    }

    #[test]
    fn with_exercised_bought_put_option() {
        let position = Position {
            price: 15.0,
            amount_owned: 100.0,
            options_bought: vec![Option {
                strike_price: 14.5,
                contract_type: OptionType::Put,
                amount: 1.0,
                price: 0.25,
            }],
            ..Position::default()
        };
        assert_eq!(calculate_expected_gain_loss(position, 13.0), -75.0);
    }

    #[test]
    fn with_no_equity_owned() {
        let position = Position {
            price: 15.0,
            amount_owned: 0.0,
            options_bought: vec![
                Option {
                    strike_price: 14.5,
                    contract_type: OptionType::Call,
                    amount: 1.0,
                    price: 0.75,
                },
                Option {
                    strike_price: 13.5,
                    contract_type: OptionType::Put,
                    amount: 1.0,
                    price: 0.05,
                },
            ],
            options_sold: vec![
                Option {
                    strike_price: 16.5,
                    contract_type: OptionType::Call,
                    amount: 1.0,
                    price: 0.5,
                },
                Option {
                    strike_price: 15.5,
                    contract_type: OptionType::Put,
                    amount: 1.0,
                    price: 0.85,
                },
            ],
        };
        assert_eq!(calculate_expected_gain_loss(position, 13.0), -145.0);
    }
}

File: app/calculate.ts
import { Position, Option, ReturnData } from './types';

// TODO: get break-even point
export function getChartPoints(position: Position): ReturnData {
  let expectedPricesOfInterest = [position.price];
  let highest = position.price,
    lowest = position.price;

  for (let bOptions of position.options_bought) {
    let chartInterest = getChartPointOfInterest(bOptions, true);
    expectedPricesOfInterest.push(bOptions.strike_price);
    expectedPricesOfInterest.push(chartInterest);

    highest = Math.max(highest, bOptions.strike_price);
    lowest = Math.min(lowest, bOptions.strike_price);
    highest = Math.max(highest, chartInterest);
    lowest = Math.min(lowest, chartInterest);
  }
  for (let sOptions of position.options_sold) {
    let chartInterest = getChartPointOfInterest(sOptions, false);
    expectedPricesOfInterest.push(sOptions.strike_price);
    expectedPricesOfInterest.push(chartInterest);

    highest = Math.max(highest, sOptions.strike_price);
    lowest = Math.min(lowest, sOptions.strike_price);
    highest = Math.max(highest, chartInterest);
    lowest = Math.min(lowest, chartInterest);
  }

  // TODO: make two additional data points before and after the range added below smarter
  expectedPricesOfInterest.push(highest + position.price * 0.05);
  expectedPricesOfInterest.push(Math.max(lowest - position.price * 0.05, 0.0));
  expectedPricesOfInterest = [...new Set(expectedPricesOfInterest)];
  expectedPricesOfInterest = expectedPricesOfInterest.sort((a, b) => a - b);

  let labels: number[] = [];
  let data: number[] = [];

  for (let expectedPrice of expectedPricesOfInterest) {
    labels.push(
      Number(
        expectedPrice.toLocaleString('en-US', {
          maximumFractionDigits: 2,
          minimumFractionDigits: 2,
          useGrouping: false,
        })
      )
    );
    data.push(calculateExpectedGainLoss(position, expectedPrice));
  }

  return {
    labels,
    data,
  };
}

function calculateExpectedGainLoss(
  position: Position,
  expected_price: number
): number {
  let costBasis = position.price * position.amount_owned;
  let amountStillOwned = position.amount_owned;

  let netPremiums: number = 0.0;
  let netOptionsProceedings: number = 0.0;
  for (let bOptions of position.options_bought) {
    netPremiums -= premiums(bOptions);
    if (shouldExercise(bOptions, expected_price)) {
      let txn_result = exercise(bOptions);
      amountStillOwned += txn_result.amount_change;
      netOptionsProceedings += txn_result.cash_change;
    }
  }
  for (let sOptions of position.options_sold) {
    netPremiums += premiums(sOptions);
    if (shouldExercise(sOptions, expected_price)) {
      let txn_result = exercise(sOptions);
      amountStillOwned -= txn_result.amount_change;
      netOptionsProceedings -= txn_result.cash_change;
    }
  }

  return (
    netPremiums +
    netOptionsProceedings +
    amountStillOwned * expected_price -
    costBasis
  );
}

function premiums(option: Option): number {
  return option.premium * option.amount * 100.0;
}

function exercise(option: Option): {
  amount_change: number;
  cash_change: number;
} {
  if (option.contract_type === 'Call') {
    return {
      amount_change: option.amount * 100.0,
      cash_change: -option.strike_price * option.amount * 100.0,
    };
  } else {
    return {
      amount_change: -option.amount * 100.0,
      cash_change: option.strike_price * option.amount * 100.0,
    };
  }
}

function shouldExercise(option: Option, expected_price: number): boolean {
  if (option.contract_type === 'Call') {
    return option.strike_price < expected_price;
  } else {
    return option.strike_price > expected_price;
  }
}

function getChartPointOfInterest(option: Option, bought: boolean): number {
  if (option.contract_type === 'Call') {
    if (bought) {
      return option.strike_price + option.premium;
    } else {
      return option.strike_price - option.premium;
    }
  } else {
    if (bought) {
      return option.strike_price - option.premium;
    } else {
      return option.strike_price + option.premium;
    }
  }
}
